### PR TITLE
Add payer balance check to FlowFees

### PR DESCRIPTION
**Pull request by janezpodhostnik**
ref: onflow/flow-go#2767

The FVM needs needs to check if the payer has sufficient balance for the transaction to be executed at all.

This PR adds the function `verifyPayersBalanceForTransactionExecution` that does this.

This function also returns the maximum possible fees the payer could incur for this transaction. This information will be needed in the storage check later.

### CODE REVIEW

File: contracts/FlowFees.cdc [hunk 1]
```diff
@@ -70,6 +71,59 @@ pub contract FlowFees {
         }
     }
 
+    // VerifyPayerBalanceResult is returned by the verifyPayersBalanceForTransactionExecution function
+    pub struct VerifyPayerBalanceResult {
+        // True if the payer has sufficient balance for the transaction execution to continue
+        pub let canExecuteTransaction: Bool
+        // The minimum payer balance required for the transaction execution to continue
+        pub let requiredBalance: UFix64
```
Comment (joshuahannan) [inline contracts/FlowFees.cdc:79]: So this balance does not take into account any potential execution effort? might be good to clarify that in the comment

File: contracts/FlowFees.cdc [hunk 2]
```diff
@@ -70,6 +71,59 @@ pub contract FlowFees {
...
+        // The maximum fees the transaction can incur (if all available execution effort is used)
+        pub let maximumTransactionFees: UFix64
```
Comment (joshuahannan) [inline contracts/FlowFees.cdc:81]: Is this the `requiredBalance + the max execution effort` or is it just the max execution effort?

File: contracts/FlowFees.cdc [hunk 3]
```diff
@@ -70,6 +71,59 @@ pub contract FlowFees {
...
+        init(canExecuteTransaction: Bool, requiredBalance: UFix64,  maximumTransactionFees: UFix64){
+            self.canExecuteTransaction = canExecuteTransaction
+            self.requiredBalance = requiredBalance
+            self.maximumTransactionFees = maximumTransactionFees
+        }
+
+    }
+
+    // verifyPayersBalanceForTransactionExecution is called by the FVM before executing a transaction.
+    // If verifies that the transaction payer's balance is high enough to continue transaction execution,
```
Comment (joshuahannan) [inline contracts/FlowFees.cdc:92]:
```suggestion
    // It verifies that the transaction payer's balance is high enough to continue transaction execution,
```

**Review by joshuahannan — APPROVED** [review]
Looks good to me! I just needed a few clarifications

File: contracts/FlowFees.cdc (hunk 1 above)
Comment (janezpodhostnik) [reply contracts/FlowFees.cdc:79]: It does. It assumes that the execution effort is the maximum possible (the computation/gas limit set by the transaction).

**Pull request merged**